src/main/scala/util/RunUtils.scala

```
@@ -119,24 +119,32 @@ object IRLoading {
     */
   def load(q: ILLoadingConfig): IRContext = {
     // TODO: this tuple is large, should be a case class
-    val (symbols, externalFunctions, globals, funcEntries, globalOffsets, mainAddress) =
-      IRLoading.loadReadELF(q.relfFile, q)
-
-    val program: Program = if (q.inputFile.endsWith(".adt")) {
-      val bapProgram = loadBAP(q.inputFile)
-      val IRTranslator = BAPToIR(bapProgram, mainAddress)
-      IRTranslator.translate
-    } else if (q.inputFile.endsWith(".gts")) {
-      loadGTIRB(q.inputFile, mainAddress)
-    } else {
-      throw Exception(s"input file name ${q.inputFile} must be an .adt or .gts file")
+
+    val withRelf = for {
+      relf <- q.relfFile
+      (symbols, externalFunctions, globals, funcEntries, globalOffsets, mainAddress) = IRLoading.loadReadELF(relf, q)
+      program: Program =
+        if (q.inputFile.endsWith(".adt")) {
+          val bapProgram = loadBAP(q.inputFile)
+          val IRTranslator = BAPToIR(bapProgram, mainAddress)
+          IRTranslator.translate
+        } else if (q.inputFile.endsWith(".gts")) {
+          loadGTIRB(q.inputFile, Some(mainAddress), Some(q.mainProcedureName))
+        } else {
+          throw Exception(s"input file name ${q.inputFile} must be an .adt or .gts file")
+        }
+      specification = IRLoading.loadSpecification(q.specFile, program, globals)
+    } yield (IRContext(symbols, externalFunctions, globals, funcEntries, globalOffsets, specification, program))
+
+    val ctx = withRelf.getOrElse {
+      val prog = loadGTIRB(q.inputFile, None, Some(q.mainProcedureName))
+      IRLoading.load(prog)
```

This seems like a particularly convoluted approach that obscures the logic. If `--relf` isn't provided, it also assumes that a `.gts` file was provided as input, which could cause confusion. If you don't want to make `BAPToIR` support returning a main procedure based on the provided name then you should at least throw an exception.